### PR TITLE
Standard library additions and cleanup.

### PR DESCRIPTION
**Pull request by gdotdesign**
This PR aims to extend, standardize and clean up the standard library modules:

- Standardized names
- Standardized tests
- Add missing functions (mainly from MDN)
- Write documentation about the conventions used

Status:
- [x] `AnimationFrame.mint`
- [x] `Array.mint`
- [x] ➕ `ArrayBuffer.mint` 
- [x] `Bool.mint`
- [x] `Clipboard.mint`
- [x] `Debug.mint`
- [x] `Dom/Dimensions.mint`
- [x] ~`Dom.mint`~ (will be a separate PR)
- [x] `File.mint`
- [x] `FileSize.mint`
- [x] `FormData.mint`
- [x] `Function.mint`
- [x] `Html/DataTransfer.mint`
- [x] `Html.Event.mint`
- [x] `Html.mint`
- [x] `Html.Portals.Body.mint`
- [x] `Html.Portals.Head.mint`
- [x] `Http.mint`
- [x] `If.mint`
- [x] `IntersectionObserver.mint`
- [x] `Json.mint`
- [x] `Map.mint`
- [x] `Math.mint`
- [x] `Maybe.mint`
- [x] `MutationObserver.mint`
- [x] `Number.mint`
- [x] `Object/Decode.mint`
- [x] `Object/Encode.mint`
- [x] `Object/Error.mint`
- [x] `Promise.mint`
- [x] `Provider/AnimationFrame.mint`
- [x] `Provider/ElementSize.mint`
- [x] `Provider/Intersection.mint`
- [x] `Provider/Keydown.mint`
- [x] `Provider/Keyup.mint`
- [x] `Provider/MediaQuery.mint`
- [x] `Provider/Mouse.mint`
- [x] `Provider/Mutation.mint`
- [x] `Provider/OutsideClick.mint`
- [x] ➕ `Provider/Pointer.mint`
- [x] `Provider/Resize.mint`
- [x] `Provider/Scroll.mint`
- [x] `Provider/Shortcuts.mint`
- [x] `Provider/TabFocus.mint`
- [x] `Provider/Tick.mint`
- [x] `Provider/Url.mint`
- [x] `Provider/Websocket.mint`
- [x] `Regexp.mint`
- [x] `ResizeObserver.mint`
- [x] `Result.mint`
- [x] `SearchParams.mint`
- [x] `Set.mint`
- [x] `Storage/Common.mint`
- [x] `Storage/Local.mint`
- [x] `Storage/Session.mint`
- [x] `String.mint`
- [x] ~`Test/Context.mint`~ (will be a separate PR for tests)
- [x] ~`Test/Html.mint`~ (will be a separate PR for tests)
- [x] ~`Test/Window.mint`~ (will be a separate PR for tests)
- [x] `Time.mint`
- [x] `Timer.mint`
- [x] `Uid.mint`
- [x] `Unless.mint`
- [x] `Url.mint`
- [x] `Validation.mint`
- [x] `WebSocket.mint`
- [x] `Window.mint`

### CODE REVIEW

File: core/source/Array.mint [hunk 1]
```diff
@@ -134,17 +134,12 @@ module Array {
...
-        return #{Maybe::Nothing}
-      }
-    })()
-    `
+    Array.first(
+      for (item of array) {
+        item
+      } when {
+        function(item)
+      })
```
Comment (Sija) [inline core/source/Array.mint:142]: As far as I understand it, this creates the array just to dispose it, seems a bit wasteful :/

File: core/source/Map.mint [hunk 2]
```diff
@@ -1,64 +1,112 @@
...
+  fun findKeyBy (
+    function : Function(value, Bool),
+    map : Map(key, value)
+  ) : Maybe(key) {
+    Array.first(
+      for (key, value of map) {
+        key
+      } when {
+        function(value)
+      })
```
Comment (Sija) [inline core/source/Map.mint:77]: ditto

File: core/source/Map.mint [hunk 3]
```diff
@@ -1,64 +1,112 @@
...
-      return #{Maybe::Nothing}
-    })()
-    `
+  fun get (search : key, map : Map(key, value)) : Maybe(value) {
+    Array.first(
+      for (key, value of map) {
+        value
+      } when {
+        key == search
+      })
```
Comment (Sija) [inline core/source/Map.mint:105]: ditto

File: core/source/Map.mint [hunk 4]
```diff
@@ -68,8 +116,70 @@ module Map {
...
+    |> Map.set("a", 1)
+    |> Map.has("a")) == true
+  */
+  fun has (search : key, map : Map(key, value)) : Bool {
+    Array.first(
+      for (key, value of map) {
+        true
+      } when {
+        key == search
+      }) or false
```
Comment (Sija) [inline core/source/Map.mint:135]: ditto

File: src/compilers/top_level.cr [hunk 5]
```diff
@@ -275,6 +275,15 @@ module Mint
           }
         }
 
+        const _o = (item, value) => {
+          if (item !== undefined &&
+              item !== null) {
```
Comment (Sija) [inline src/compilers/top_level.cr:280]:
```suggestion
          if (item !== undefined && item !== null) {
```

File: core/source/Array.mint (hunk 1 above)
Comment (gdotdesign) [reply core/source/Array.mint:142]: The intention here (and the others like it) was to rewrite native (JavaScript) code to Mint.

File: src/ls/code_actions/provider_actions.cr [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  module LS
+    class CodeAction < LSP::RequestMessage
+      module ProviderActions
+        extend self
+
+        def order_entities(node, workspace, uri)
+          # Save the original order of constants and functions
+          order =
+            (node.functions + node.constants + node.states + node.gets)
+              .sort_by(&.from)
```
Comment (Sija) [inline src/ls/code_actions/provider_actions.cr:11]:
```suggestion
              .sort_by!(&.from)
```

**Review by Sija — APPROVED** [review]
One minor comment, LGTM in overall

**Pull request merged**